### PR TITLE
sasquatch: init

### LINKED ISSUE #66270: add sasquatch
**Project description**

sasquatch is a varation of the unsquashfs utilty that can unpack a much larger amount of vendor specific flavours of squashfs that the regular unsquashfs utility cannot unpack.

It is distributed as a patch to the regular squashfs-tools package (so might need some extra work to work with the squashfs-tools package already in nixpkgs).

**Metadata**

* homepage URL: https://github.com/devttys0/sasquatch
* source URL: https://github.com/devttys0/sasquatch
* license: gpl2+
* platforms: linux

### PR DESCRIPTION
**Pull request by Pamplemousse**
Fixes #66270 .

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/filesystems/sasquatch/default.nix [hunk 1]
```diff
@@ -0,0 +1,50 @@
+{ fetchFromGitHub
+, fetchurl
+, lib
+, lz4 ? null
+, lz4Support ? false
+, lzma
+, lzo
+, stdenv
+, wget
+, xz
```
Comment (aanderse) [inline pkgs/tools/filesystems/sasquatch/default.nix]: I think you meant to add this to `buildInputs`.

File: pkgs/tools/filesystems/sasquatch/default.nix [hunk 2]
```diff
@@ -0,0 +1,50 @@
+{ fetchFromGitHub
+, fetchurl
+, lib
```
Comment (aanderse) [inline pkgs/tools/filesystems/sasquatch/default.nix]: With changes I mention below you shouldn`t need `lib` anymore.

File: pkgs/tools/filesystems/sasquatch/default.nix [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    owner = "devttys0";
+    repo = "sasquatch";
+    rev = "3e0cc40fc6dbe32bd3a5e6c553b3320d5d91ceed";
+    sha256 = "19lhndjv7v9w6nmszry63zh5rqii9v7wvsbpc2n6q606hyz955g2";
+  } + "/patches/patch0.txt";
+in
+stdenv.mkDerivation rec {
+  name = "sasquatch";
+
+  src = fetchurl {
```
Comment (aanderse) [inline pkgs/tools/filesystems/sasquatch/default.nix]: Does this work with the version of `squashfs` we have in nixpkgs (`4.4dev_20180612`)?

File: pkgs/tools/filesystems/sasquatch/default.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    url = "https://downloads.sourceforge.net/project/squashfs/squashfs/squashfs4.3/squashfs4.3.tar.gz";
+    sha256 = "1xpklm0y43nd9i6jw43y2xh5zvlmj9ar2rvknh0bh7kv8c95aq0d";
+  };
+
+  buildInputs = [ lzma lzo wget zlib ];
```
Comment (aanderse) [inline pkgs/tools/filesystems/sasquatch/default.nix]: Don't forget to add `stdenv.lib.optional lz4Support lz4;` onto `buildInputs`.

File: pkgs/tools/filesystems/sasquatch/default.nix [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+, lz4 ? null
+, lz4Support ? false
+, lzma
+, lzo
+, stdenv
+, wget
+, xz
+, zlib
+}:
+
```
Comment (aanderse) [inline pkgs/tools/filesystems/sasquatch/default.nix]: You should add `assert lz4Support -> (lz4 != null);`.

**Review by aanderse — CHANGES_REQUESTED** [review]
Hi! Thanks for deciding to contribute to nixpkgs and adding yourself as a maintainer. Sorry we took a while to get around to reviewing your PR.

I have made some comments I hope you will find helpful. The most important comment is my question about whether the version of `squashfs` we have in nixpkgs is compatible with the patch `sasquatch` provides. If the patch is compatible we could just modify the existing `squashfs` package to make a new `sasquatch` variant of it, and save yourself a fair bit of work. Let me know.

Please let me know if you have any questions.

**Review by aanderse — CHANGES_REQUESTED** [review]
I also noticed `NIX_LDFLAGS = "-lgcc_s"; # for pthread_cancel` from the history of our [squashfsTools](https://github.com/NixOS/nixpkgs/blob/56fa7ec87f5ba8658bea3635e2fb22785832cb29/pkgs/tools/filesystems/squashfs/default.nix). Is that needed?

And how about CVEs listed against `squashfsTools` at version 4.3? Any?

File: maintainers/maintainer-list.nix [hunk 6]
```diff
@@ -3900,6 +3900,11 @@
     github = "pakhfn";
     name = "Fedor Pakhomov";
   };
+  pamplemousse = {
+    email = "xav.maso@gmail.com";
+    github = "Pamplemousse";
```
Comment (aanderse) [inline maintainers/maintainer-list.nix]: As per brand new policy if you set `github` you must now also set `githubid`. Your `githubid` is [2647236](https://api.github.com/users/pamplemousse).

File: pkgs/tools/filesystems/sasquatch/default.nix [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    owner = "devttys0";
+    repo = "sasquatch";
+    rev = "3e0cc40fc6dbe32bd3a5e6c553b3320d5d91ceed";
+    sha256 = "19lhndjv7v9w6nmszry63zh5rqii9v7wvsbpc2n6q606hyz955g2";
+  } + "/patches/patch0.txt";
+in
+stdenv.mkDerivation rec {
+  pname = "sasquatch";
+  version = "4.3";
+  name = "${pname}-${version}";
```
Comment (aanderse) [inline pkgs/tools/filesystems/sasquatch/default.nix]: How did that sneak in? If you set `pname` and `version` then you don't need to set `name` as `name = "${pname}-${version};` is already implied.

File: pkgs/tools/filesystems/sasquatch/default.nix [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  '';
+
+  installFlags = "INSTALL_DIR=\${out}/bin";
+
+  makeFlags = [ "XZ_SUPPORT=1" ]
+    ++ stdenv.lib.optional lz4Support "LZ4_SUPPORT=1";
+
+  meta = with stdenv.lib; {
+    homepage = "https://github.com/devttys0/sasquatch";
+    description = "Set of patches to the standard unsquashfs utility (part of squashfs-tools) that attempts to add support for as many hacked-up vendor-specific SquashFS implementations as possible.";
```
Comment (aanderse) [inline pkgs/tools/filesystems/sasquatch/default.nix]: Please omit period at end of `description`.

**Review by aanderse — CHANGES_REQUESTED** [review]
Thanks for following up on the `license`! Just a few small things I noticed.

**Comment by Pamplemousse** [comment]
Regarding your comment about the `-lgcc_s` flag, I do not know.

I had a look around:
  * The `squashfs` patch does not seem to touch anything related to `pthread_cancel`;
  * The used `Makefile` (from squashfs, used in [`sasquatch`'s build script](https://github.com/devttys0/sasquatch/blob/master/build.sh#L37)), does not have any kind of mention of the `gcc_s` lib.

So, I would not be inclined adding it.

**Comment by Pamplemousse** [comment]
Regarding the CVEs, I have can found two related to `sasquatch`: [CVE-2015-4645](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2015-4645) and [CVE-2015-4646](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2015-4646).

As per [this comment on devttys0/sasquatch#5](https://github.com/devttys0/sasquatch/pull/5#issuecomment-390988066), fixes are taken care of upstream (in `squashfs`).

**Comment by aanderse** [comment]
Thanks. Maybe @ruuda, @danbst, or @charles-dyfis-net could take a quick look as people familiar with `squashfsTools`.

**Review by aanderse — APPROVED** [review]

**Comment by ruuda** [comment]
Looks good to me. I think it is good to keep the sources separate and not base this package on `squashfstools`. Otherwise, when there is a `squashfstools` update and the patches no longer apply, we either need to wait for upstream Sasquatch to update the patches before we can upgrade `squashfstools`, or we need to disable the `sasquatch` package. Now they can coexist.

**Comment by aanderse** [comment]
@Pamplemousse at this point the only thing missing is a license then. Ready for merge once that situation is resolved. Thanks! :+1:

**Comment by aanderse** [comment]
Thanks @Pamplemousse, and welcome to the team :tada:

**Pull request merged**